### PR TITLE
Fastify plugin

### PR DESCRIPTION
**Pull request by delvedor**
As we have discussed in https://github.com/fastify/fastify/issues/1503 and https://github.com/fastify/fastify/pull/2014, Middie has been refactored to be a Fastify plugin that adds the same functionality that Fastify was providing in v2 and v1.
The Middie engine itself can be still used without the plugin code by requiring it with `middie/engine`.

*This is a semver-major change.*

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,121 @@
...
 The syntax style is the same as [express](http://npm.im/express)/[connect](https://www.npmjs.com/package/connect).
 Does not support the full syntax `middleware(err, req, res, next)`, because error handling is done inside Fastify.
 
-If you want to see how use this module with Fastify, check [here](https://github.com/fastify/fastify/#fastifyusemiddlewarereq-res-next).
-
 ## Install
 
 ```
-npm install middie --save
+npm install middie
```
Comment (RafaelGSS) [inline README.md]: why not `--save`?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  fastify.use = use
+  fastify[kMiddlewares] = []
+  fastify[kMiddie] = Middie(onMiddieEnd)
+
+  fastify
+    .addHook('onRequest', enhanceRequest)
+    .addHook('onRequest', runMiddie)
+    .addHook('onRegister', onRegister)
+
+  function use (path, fn) {
```
Comment (RafaelGSS) [inline index.js]: In examples are using multiples paths: `['css/*', 'js/*']`. I think that here is not covering it. Right?

File: README.md (hunk 1 above)
Comment (delvedor) [reply README.md]: Because from npm@5 is the default operation ;)

File: index.js (hunk 2 above)
Comment (delvedor) [reply index.js]: It should be covered in the Middie engine tests.

**Comment by Eomm** [comment]
What do you expect if the middleware consumes the request via `.end()`?

**Comment by davidmarkclements** [comment]
> I agree, but this a long-standing issue, fixing it in avvio is quite hard since it's designed to avoid this behavior and add it in Fastify means "trick" avvio.

@delvedor fixed it 🎉 https://github.com/mcollina/avvio/pull/92

File: index.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    for (const middleware of middlewares) {
+      instance.use(...middleware)
+    }
+  }
+
+  next()
+}
+
+module.exports = fp(middiePlugin, {
+  // fastify: '>=3.0.0',
```
Comment (Eomm) [inline index.js]:
```suggestion
  // TODO fastify: '>=3.0.0',
```

File: README.md [hunk 4]
```diff
@@ -1,25 +1,121 @@
...
 ```
-<a name="usage"></a>
+
 ## Usage
+Register the plugin and start using your middlewares.
+```js
+const fastify = require('fastify')()
+fastify.register(require('middie'))
+  .after(() => {
+     // do you know we also have cors support?
```
Comment (mcollina) [inline README.md]: please use the new await app.after.

Comment (mcollina) [reply README.md:22]: change this to await fastify.register()

**Pull request merged**